common/grammar-parser.cpp

```diff
@@ -174,11 +174,11 @@ namespace grammar_parser {
             //            Sstar  ::= Scopy Sstar |
 
             uint32_t content_rule_id = 0;
-            if (last_sym_start == out_elements.size() - 1) {
+            if (last_sym_start >= 0 && last_sym_start == out_elements.size() - 1 && out_elements[last_sym_start].type == LLAMA_GRETYPE_RULE_REF) {
```

Comment:
I'm trying to wrap my head around this change, and its potential impact -- best I can tell, this is the change that you refer to in the PR description re:

> Except if S was trivial at the first place (e.g. single ref), then S is not 'copied' to a distinct rule, e.g.:
> ```
> S+ --> S'     ::= S Sstar
>        Sstar  ::= S Sstar |
> ```
> (this 'copy' mechanism is to avoid repeating long S sequences - think large groups, or complex char ranges)

One thing I haven't fully wrapped my head around is if the new rewrite rules are more efficient than they were before, less efficient, or about the same. I might need to expand the gbnf-validator and compare the parsed grammars between old-and-new before I really wrap my mind around what's going on here.

Reply:
> if the new rewrite rules are more efficient than they were before, less efficient, or about the same 

Yeah only some benchmarking will tell (and we’d need more real-world grammars for that). My hunch is it’s probably the same, the extra indirection might be negligible, and the saving of avoiding duplication negligible.